Mathlib/Data/Quot.lean

```diff
@@ -414,9 +414,25 @@ instance piSetoid {ι : Sort _} {α : ι → Sort _} [∀ i, Setoid (α i)] : Se
             fun h _ ↦ Setoid.symm (h _),
             fun h₁ h₂ _ ↦ Setoid.trans (h₁ _) (h₂ _)⟩
 
-/-- Given a function `f : Π i, Quotient (S i)`, returns the class of functions `Π i, α i` sending
-each `i` to an element of the class `f i`. -/
-noncomputable def Quotient.choice {ι : Type _} {α : ι → Type _} [S : ∀ i, Setoid (α i)]
+/-- Given a class of functions `q : @Quotient (∀ i, α i) _`, returns the class of `i`-th projection
+`Quotient (S i)`. -/
+def Quotient.proj {ι : Type _} {α : ι → Type _} [S : ∀ i, Setoid (α i)]
+    (q : @Quotient (∀ i, α i) (by infer_instance)) (i) :
+    Quotient (S i) :=
+  q.map (· i) (fun _ _ h ↦ by exact h i)
+
+private unsafe def Quotient.choiceImpl {ι : Type _} {α : ι → Type _} [S : ∀ i, Setoid (α i)] :
+    (∀ i, Quotient (S i)) → @Quotient (∀ i, α i) (by infer_instance) :=
+  unsafeCast
+
+/-- Given a function `f : ∀ i, Quotient (S i)`, returns the class of functions `∀ i, α i` sending
+each `i` to an element of the class `f i`.
+
+This function uses `Classical.choice` to construct data, but it has special support in the runtime.
+For the version for quotients indexed by a finite type that does not depend on `Classical.choice`,
+see `Quotient.finChoice`. -/
+@[implemented_by choiceImpl]
```

Comment:
```suggestion
```

This suggestion comes up from time to time, but it's inconsistent to make the function computable.  (Inconsistent in the sense that you can write a function which provably returns true, but evaluates to false in the VM.)  See this [example from Chris Hughes](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/quotients.20and.20products.20again/near/187614136):

```lean
import Mathlib.Data.Setoid.Basic

def exoticInstance : Decidable True :=
  -- ⊤ is the always true relation
  @Quot.recOnSubsingleton _ _ (fun _ => Decidable True) _
    (@Quotient.choice (@Quotient Bool ⊤) (fun _ => Bool) ⊤ id)
    (fun f => decidable_of_iff (f ⟦false⟧ = f ⟦true⟧)
      (iff_true_intro (congr_arg f (Quotient.sound trivial))))

-- VM disagrees with type theory:
example : @decide True exoticInstance = true  := .trans (by congr) decide_True
example : @decide True exoticInstance = false := by native_decide

#eval exoticInstance -- isFalse _
```

Reply:
😮